test_conformance/printf/util_printf.cpp

```diff
@@ -859,16 +786,130 @@ size_t verifyOutputBuffer(char *analysisBuffer,testCase* pTestCase,size_t testId
             ++exp;
+        while(*eCorrectBuffer == '0')
+            ++eCorrectBuffer;
         return strcmp(eCorrectBuffer,exp);
     }
-    if(!strcmp(pTestCase->_correctBuffer[testId],"inf"))
+    if(!strcmp(pTestCase->_correctBuffer[testId].c_str(),"inf"))
     return strcmp(analysisBuffer,"inf")&&strcmp(analysisBuffer,"infinity")&&strcmp(analysisBuffer,"1.#INF00")&&strcmp(analysisBuffer,"Inf");
-    if(!strcmp(pTestCase->_correctBuffer[testId],"nan") || !strcmp(pTestCase->_correctBuffer[testId],"-nan")) {
+    if(!strcmp(pTestCase->_correctBuffer[testId].c_str(),"nan") || !strcmp(pTestCase->_correctBuffer[testId].c_str(),"-nan")) {
        return strcmp(analysisBuffer,"nan")&&strcmp(analysisBuffer,"-nan")&&strcmp(analysisBuffer,"1.#IND00")&&strcmp(analysisBuffer,"-1.#IND00")&&strcmp(analysisBuffer,"NaN")&&strcmp(analysisBuffer,"nan(ind)")&&strcmp(analysisBuffer,"nan(snan)")&&strcmp(analysisBuffer,"-nan(ind)");
     }
-    return strcmp(analysisBuffer,pTestCase->_correctBuffer[testId]);
+    return strcmp(analysisBuffer,pTestCase->_correctBuffer[testId].c_str());
+}
+
+static void intRefBuilder(printDataGenParameters& params)
+{
+    printf(params.genericFormat, atoi(params.dataRepresentation));
+}
+
+static void floatRefBuilder(printDataGenParameters& params)
+{
+    printf(params.genericFormat, strtof(params.dataRepresentation, NULL));
+}
+
+static void octalRefBuilder(printDataGenParameters& params)
+{
+    const unsigned long int data = strtoul(params.dataRepresentation, NULL, 10);
+    printf(params.genericFormat, data);
+}
+
+static void unsignedRefBuilder(printDataGenParameters& params)
+{
+    const unsigned long int data = strtoul(params.dataRepresentation, NULL, 10);
+    printf(params.genericFormat, data);
+}
+
+static void hexRefBuilder(printDataGenParameters& params)
+{
+    const unsigned long int data = strtoul(params.dataRepresentation, NULL, 0);
+    printf(params.genericFormat, data);
 }
 
+/*
+    Generate reference results.
+
+    Results are only generated for test cases
+    that can easily be generated by using CPU
+    printf.
+
+    If that is not the case, results are constants
+    that have been hard-coded.
+*/
+void generateRef(const cl_device_id device)
+{
+    int fd = -1;
+    char _refBuffer[ANALYSIS_BUFFER_SIZE];
+    cl_device_fp_config fpConfig;
+    const RoundingMode hostRound = get_round();
+    RoundingMode deviceRound;
+
+    // Discover device rounding mode
+    cl_int err = clGetDeviceInfo(device,
+                                CL_DEVICE_SINGLE_FP_CONFIG,
+                                sizeof(fpConfig),
+                                &fpConfig,
+                                NULL);
+    if (err != CL_SUCCESS)
+    {
+        log_error("clGetDeviceInfo failed");
+        return;
+    }
+
+    // Map device rounding to CTS rounding type
+    if (fpConfig & CL_FP_ROUND_TO_NEAREST)
+    {
+        deviceRound = kRoundToNearestEven;
+    }
+    else if (fpConfig & CL_FP_ROUND_TO_ZERO)
```

Comment:
What if an implementation reports several supported rounding modes? This isn't forbidden by the spec. There doesn't seem to be a way for the implementation to report which is the default though.

Reply:
Yes this is a tricky one. Without a way of knowing which mode will actually be used by the implementation, I'm not sure we can solve this at the moment (aside from attempting to detect the mode in-kernel like bruteforce does).

Reply:
There is no way to solve this properly without changing the spec I think. This approach works because it covers the most common case (an embedded profile that won't support RTE but will support RTZ). It's essentially the same logic as `get_default_rounding_mode` and I think we probably ought to be using that to limit the number of places where we hard-code that assumption.

I suggest you create a Gitlab issue to add a query to report the default rounding mode or at least discuss this. We should definitely make all of this a lot clearer. What does it mean to support multiple rounding modes for example if there can be only one default? Can the default rouding mode change in some circumstances, etc? :).